--- a/helix-term/src/application.rs
+++ b/helix-term/src/application.rs
@@ -633,7 +633,36 @@ pub async fn handle_language_server_message(
                                         source: diagnostic.source.clone()
                                     })
                                 })
-                                .collect();
+                                .collect::<Vec<_>>();
+
+                            doc.clear_text_annotations("diagnostics");
+
+                            use helix_core::diagnostic::Severity;
+                            use helix_view::decorations::{TextAnnotation, TextAnnotationKind};
+                            use helix_view::graphics::{Color, Style};
+                            let style = Style::default().bg(Color::Gray);
+                            doc.push_text_annotations(
+                                "diagnostics",
+                                diagnostics
+                                    .iter()
+                                    .map(|d| TextAnnotation {
+                                        text: d.message.clone().into(),
+                                        style: d
+                                            .severity
+                                            .as_ref()
+                                            .map(|s| match s {
+                                                Severity::Error => style.clone().fg(Color::Red),
+                                                Severity::Warning => {
+                                                    style.clone().fg(Color::Yellow)
+                                                }
+                                                Severity::Info => style.clone().fg(Color::Blue),
+                                                Severity::Hint => style.clone().fg(Color::Green),
+                                            })
+                                            .unwrap_or(style.clone().fg(Color::Yellow)),
+                                        line: d.line,
+                                        kind: TextAnnotationKind::Eol,
+                                    })
+                            );

                             doc.set_diagnostics(diagnostics);
                         }
